File: 프로그래머스/2/72412. 순위 검색/순위 검색.cpp
#include <string>
#include <vector>
#include <iostream> 
#include <unordered_map> 
#include <sstream>
#include <algorithm>
using namespace std;

unordered_map<string, vector<int>> scores;

// 벡터에서 해당 score 이상인 지원자 수 리턴
int binary_search(vector<int> & scores, int score){
    int left = 0; int right = scores.size()-1;
    int point = scores.size(); 
    while(left <= right){
        int mid = (right + left)/2;
        if(scores[mid] >= score) {
            point = mid;
            right = mid - 1;
        }
        else left = mid + 1;
    }
    // cout << "p: " << point << endl;
    return scores.size() - point;
}

vector<int> solution(vector<string> info, vector<string> query) {
    vector<int> answer;
    
    // 24가지 map 중 한곳에 넣기 (5만*5)
    for(int i = 0; i < info.size(); i++){
        string content = info[i]; 
        stringstream ss(content);
        string value; 
        string key = "";
        for(int i = 0; i < 4; i++){
            ss >> value; 
            key += value; // 언어, 직군, 경력, 소울푸드
            key += "_";
        }
        ss >> value;
        scores[key].push_back(stoi(value));   
        // cout << key << " " << value << endl;
    }
    
    // 각 벡터 정렬(오름차순)
    for(auto & score : scores){
        sort(score.second.begin(), score.second.end());
    }
    
    // 각 조건에 따라 벡터 찾고 이진탐색으로 개수 찾기 (10만 * log(5만) * 24)
    for(int i = 0; i < query.size(); i++){ // 10만
        string content = query[i]; 
        stringstream ss(content);
        string value, score; 
        vector<string> q;
        for(int i = 0; i < 7; i++){
            ss >> value; 
            if(value != "-" && value != "and" && value != " ") q.push_back(value);
        }
        ss >> score;

        // 적합한 key 찾기
        vector<string> keys; 
        for(auto & m: scores){ // 24
            string key = m.first; 
            bool fit = true;
            for(int i = 0; i < q.size(); i++){ // 4 쿼리의 모든 조건에 맞아야 함. 
                // cout << key << " " << q[i] << endl;
                if(key.find(q[i]) == string::npos) {
                    fit = false; 
                    // cout << "false" << endl;
                    break;
                }
            }
            if(fit) {
                keys.push_back(key);
                // cout << key << endl;
            }
        }
        
        // key 순회하며 이진탐색으로 개수 찾아 합하기
        int sum = 0; 
        for(auto & k : keys){
            // cout << k << " " << score << endl;
            sum += binary_search(scores[k], stoi(score));
        }
        
        answer.push_back(sum); 
    }
    
    
    return answer; // 각 쿼리에 해당하는 인원 수
}

File: 프로그래머스/2/72412. 순위 검색/순위 검색.java
import java.util.*;

class Solution {
    Map<String, List<Integer>> scores = new HashMap<>();

    // 리스트에서 해당 점수 이상인 지원자 수 반환
    public int binarySearch(List<Integer> scores, int targetScore) {
        int left = 0;
        int right = scores.size() - 1;
        int point = scores.size();

        while (left <= right) {
            int mid = (left + right) / 2;

            if (scores.get(mid) >= targetScore) {
                point = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return scores.size() - point;
    }

    public int[] solution(String[] info, String[] query) {
        List<Integer> answer = new ArrayList<>();
        scores.clear();

        // 지원자 정보를 24가지 조건 중 하나의 Map에 저장
        for (String content : info) {
            String[] values = content.split(" ");

            String key = values[0] + "_"
                       + values[1] + "_"
                       + values[2] + "_"
                       + values[3] + "_";

            int score = Integer.parseInt(values[4]);

            scores.computeIfAbsent(key, ignored -> new ArrayList<>())
                  .add(score);
        }

        // 각 조건에 저장된 점수를 오름차순 정렬
        for (List<Integer> scoreList : scores.values()) {
            Collections.sort(scoreList);
        }

        // 각 쿼리에 맞는 조건을 찾고 이진 탐색으로 인원수 계산
        for (String content : query) {
            String[] values = content.split(" ");

            List<String> conditions = new ArrayList<>();

            // 쿼리 형식:
            // 언어 and 직군 and 경력 and 소울푸드 점수
            for (int i = 0; i <= 6; i += 2) {
                if (!values[i].equals("-")) {
                    conditions.add(values[i]);
                }
            }

            int targetScore = Integer.parseInt(values[7]);
            int count = 0;

            // 적합한 key 찾기
            for (Map.Entry<String, List<Integer>> entry : scores.entrySet()) {
                String key = entry.getKey();
                boolean fit = true;

                for (String condition : conditions) {
                    if (!key.contains(condition)) {
                        fit = false;
                        break;
                    }
                }

                if (fit) {
                    count += binarySearch(entry.getValue(), targetScore);
                }
            }

            answer.add(count);
        }

        // 각 쿼리에 해당하는 인원 수 반환
        return answer.stream()
                     .mapToInt(Integer::intValue)
                     .toArray();
    }
}